src/Encamina.Enmarcha.SemanticKernel/MemoryStoreExtender.cs

```diff
@@ -0,0 +1,165 @@
+    {
+        var chunkSize = await GetChunkSize(memoryId, collectionName, cancellationToken);
+        if (chunkSize > 0)
+        {
+            var keys = Enumerable.Range(0, chunkSize).Select(i => BuildMemoryIdentifier(memoryId, i));
+            await MemoryStore.RemoveBatchAsync(collectionName, keys, cancellationToken);
+            RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.RemoveBatch, Keys = keys, CollectionName = collectionName });
+        }
+    }
+
+    /// <inheritdoc/>
+    public virtual async Task<MemoryContent> GetMemoryAsync(string memoryId, string collectionName, CancellationToken cancellationToken)
+    {
+        var chunkSize = await GetChunkSize(memoryId, collectionName, cancellationToken);
+
+        if (chunkSize == 0)
+        {
+            return null;
+        }
+
+        var keys = Enumerable.Range(0, chunkSize).Select(i => BuildMemoryIdentifier(memoryId, i));
+        var memoryRecords = await MemoryStore.GetBatchAsync(collectionName, keys, cancellationToken: cancellationToken).ToListAsync(cancellationToken);
+        RaiseMemoryStoreEvent(new() { EventType = MemoryStoreEventTypes.GetBatch, Keys = keys, CollectionName = collectionName });
+
+        return new MemoryContent
+        {
+            Metadata = JsonSerializer.Deserialize<Dictionary<string, string>>(memoryRecords[0].Metadata.AdditionalMetadata),
+            Chunks = memoryRecords.Select(m => m.Metadata.Text),
+        };
+    }
+
+    /// <inheritdoc/>
+    public virtual async IAsyncEnumerable<string> BatchUpsertMemoriesAsync(string collectionName, IDictionary<string, MemoryContent> memoryContents, Kernel kernel, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var memoryRecords = new Collection<MemoryRecord>();
+        var memoryIds = new List<string>();
+
+        foreach (var memoryContentItem in memoryContents)
+        {
+            var memoryContentId = memoryContentItem.Key;
+            var memoryContent = memoryContentItem.Value;
+            var totalChunks = memoryContent.Chunks.Count();
+
+            if (totalChunks > 0)
+            {
+                memoryContent.Metadata.Add(Constants.MetadataTotalChunksCount, totalChunks.ToString());
+
+                for (var i = 0; i < totalChunks; i++)
+                {
+                    var chunk = memoryContent.Chunks.ElementAt(i);
+                    var embedding = await kernel.GetRequiredService<ITextEmbeddingGenerationService>().GenerateEmbeddingAsync(chunk, kernel, cancellationToken);
+                    memoryRecords.Add(MemoryRecord.LocalRecord($@"{memoryContentId}-{i}", chunk, null, embedding, JsonSerializer.Serialize(memoryContent.Metadata)));
+                }
+            }
+        }
+
+        var asyncKeys = MemoryStore.UpsertBatchAsync(collectionName, memoryRecords, cancellationToken);
+        var keys = await asyncKeys.ToListAsync();
```

Comment:
Let's review this code here. There are some issues with double enumerations and asynchronicity.

Reply:
Thank you for your help in reviewing this part of the code.